src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/RuntimeHelpers.cs

```diff
@@ -13,22 +14,39 @@ public static partial class RuntimeHelpers
         public delegate void CleanupCode(object userData, bool exceptionThrown);
 
         /// <summary>
-        /// GetSubArray helper method for the compiler to slice an array using a range.
+        /// Slices the specified array using the specified range.
         /// </summary>
         public static T[] GetSubArray<T>(T[] array, Range range)
         {
-            Type elementType = array.GetType().GetElementType();
-            Span<T> source = array.AsSpan(range);
+            if (array == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.array);
+            }
+
+            Range.OffsetAndLength offLen = range.GetOffsetAndLength(array.Length);
 
-            if (elementType.IsValueType)
+            if (default(T) != null || typeof(T[]) == array.GetType())
             {
-                return source.ToArray();
+                // We know the type of the array to be exactly T[].
+
+                if (offLen.Length == 0)
+                {
+                    return Array.Empty<T>();
+                }
+
+                var dest = new T[offLen.Length];
+                Buffer.Memmove(
```

Comment:
Would it make sense to operate on `byte` (i.e. omit the reinterpret-cast to `T`) as `Buffer.Memmove<T>` goes back to `byte` -> https://github.com/dotnet/coreclr/blob/13421bf4d01816fb24b974695acb9a9698193575/src/System.Private.CoreLib/shared/System/Buffer.cs#L318-L321

Reply:
I'd prefer this not have to deal with the math around element size and length.